vowpalwabbit/reductions/epsilon_decay.h

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include "array_parameters.h"
+#include "distributionally_robust.h"
+#include "reductions_fwd.h"
+#include "scored_config.h"
+
+namespace VW
+{
+namespace epsilon_decay
+{
+VW::LEARNER::base_learner* epsilon_decay_setup(VW::setup_base_i&);
+
+struct epsilon_decay_score : scored_config
+{
+  epsilon_decay_score() = default;
+  epsilon_decay_score(double alpha, double tau, uint64_t model_idx)
+      : VW::scored_config(alpha, tau), _model_idx(model_idx)
+  {
+  }
+  float get_upper_bound() const { return this->current_ips(); }
+  float get_lower_bound() const { return _lower_bound; }
+  uint64_t get_model_idx() const { return _model_idx; }
+  void update_bounds(float w, float r);
+
+  float _lower_bound = 0.f;
+  uint64_t _model_idx;
+};
+
+struct epsilon_decay_data
+{
+  epsilon_decay_data(uint64_t num_configs, uint64_t min_scope, double epsilon_decay_alpha, double epsilon_decay_tau,
+      parameters& weights)
+      : min_scope(min_scope)
+      , epsilon_decay_alpha(epsilon_decay_alpha)
+      , epsilon_decay_tau(epsilon_decay_tau)
+      , weights(weights)
+  {
+    for (uint64_t i = 0; i < num_configs; ++i)
+    {
+      epsilon_decay_score s(epsilon_decay_alpha, epsilon_decay_tau, i);
+      scored_configs.push_back(s);
+    }
+  }
+  std::vector<epsilon_decay_score> scored_configs;
+  uint64_t min_scope;
+  double epsilon_decay_alpha;
+  double epsilon_decay_tau;
+  parameters& weights;
+};
+
+}  // namespace epsilon_decay
+
+namespace model_utils
+{
+size_t read_model_field(io_buf&, VW::epsilon_decay::epsilon_decay_score&);
```

Comment:
What are these needed for? If they are only required for the save_load of this reduction then they should be able to be only in the cc file and we should avoid exporting them here

Reply:
This actually doesn't work. If I move these to only the cc file, the save_load for scored_config will be used instead of epsilon_decay_score (which inherits the former), since a vector of these are being saved, and it would not be visible to model_utils.h

Reply:
Ooof that's not good... I wonder if there is a way to ensure the parent is not considered for overload resolution and only the concrete type. I can't think of a way though